### PR TITLE
Support containerd-based graphdriver

### PR DESCRIPTION
**Pull request by ktock**
Containerd has a pluggable architecture. Some OSS projects leverage this feature and extend it with their plugins. Especially snapshotter and content store are supported as external binary plugins and this enables widening use-cases of the runtime with lower maintaining costs.

Unfortunately, these plugins can't be used on docker because it uses graphdrivers for contents and snapshots management instead of containerd plugins. This leads to duplication of work for plugin implementers towards using their custom plugins on docker in addition to containerd.

This commit solves this issue by supporting containerd's snapshotter and content store as a graphdriver on docker.

This is related to #38043 and #38738 but I separately opened this because the scope of this PR is a bit narrower. This PR is a draft because we need to cover more edge cases and needs more tests and further optimization but this should be enough as a blueprint for the discussion.

The following is an overview of this integration.

## containerd-based graphdriver

Graphdriver manages diff contents and layer snapshots (mount points). These functionalities can be mapped to content store and snapshotter in containerd. This commit includes a graphdriver based on these plugins.

Diff content management (related to `DiffDriver` APIs) is backed by the content store. Contents provided through `ApplyDiff` are stored in it so graphdriver can provide exactly same tarball through `Diff` API for these applied contents.

Layer snapshots management (related to `ProtoDriver` APIs) is provided by snapshotter. A snapshot is `Prepare`-ed when `Create` or `CreateReadWrite` API is invoked. It's committed when it needs to be a parent for new layers. Mount points are provided through snapshotter's `View` or `Mount` APIs.

## Remote snapshotters

This commit also includes support for containerd's "remote snapshotters". They allow runtimes to use remotely mounted container rootfs without pulling images, which leads to speeding up container startup especially in cold-start cases. For more details, please refer to the https://github.com/containerd/stargz-snapshotter repo which is one of the implementations of remote snapshotters. Benchmarking result on containerd with stargz snapshotter shows performance improvement on container startup (not measured on docker yet but improvements can be expected also on it).

The benchmarking result on 288c338

The graphdriver in this patch supports remote snapshotters. Layer snapshots and diff contents can be provided without pulling images. I call it "remote graphdriver" here.

Remote graphdriver allows clients (including docker daemon core) to check if a layer exists in the graphdriver. In this patch, this checking is done through `Create` and `CreateReadWrite` APIs during `docker pull` and if any layers in the targeting image exist in graphdriver, docker skips the pull for these layers. This patch introduces `TargetOpts` options passed via `CraeteOpts` for specifying the targeting layer.

```go
type CreateOpts struct {
...
	TargetInfo *TargetOpts
}

type TargetOpts struct {
	MountChainID  digest.Digest
	ContentDiffID digest.Digest
	ImageRef      string
	LayerDigest   digest.Digest
	ImageLayers   []digest.Digest
}
```

If the snapshots (specified by ChainID) and diff contents (specified by DiffID) exist in the backing snapshotter and content store (which is backed by remote stores), the graphdriver returns `ErrAlreadyExists` to tell clients the layer existence.

In the current implementation, the layer validation (means layer provided by graphdriver has ChainID for its mount point and has DiffID for its inflated diff content) is snapshotter's responsibility but we might need more discussion for it.

The example implementation of a snapshotter which supports "remote graphdriver" is on https://github.com/ktock/stargz-snapshotter/tree/graphdriver which is a patched version of containerd/stargz-snapshotter.

## Questions

I'm keen to move the containerd-docker integration forward and any feedback and comments are very welcome.

- Could I get your comments on this design and implementation? Is the kind of "containerd-based graphdriver" acceptable? If not, are there any considerations or alternatives?
- I contained both "containerd-based graphdriver" and "remote graphdriver" in this single PR because both of them are closely related. Please tell me if I should make separated PRs.
- Please tell me anything missing in this design and implementation.

### CODE REVIEW

File: daemon/graphdriver/snapshotter/snapshotter.go [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+	"github.com/docker/docker/pkg/stringid"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	driverName = "snapshotter"
```
Comment (AkihiroSuda) [inline daemon/graphdriver/snapshotter/snapshotter.go:38]: I prefer "containerd"

Comment (AkihiroSuda) [reply daemon/graphdriver/snapshotter/snapshotter.go:38]: also NS should be "moby"

Comment (ktock) [reply daemon/graphdriver/snapshotter/snapshotter.go]: Thanks for the feedback. I renamed the graphdriver to "containerd" and fixed it to use "moby" namespace.

File: distribution/xfer/download.go [hunk 2]
```diff
@@ -89,6 +91,12 @@ type DownloadDescriptor interface {
 	Close()
 }
 
+type DownloadAddressableDescriptor interface {
```
Comment (AkihiroSuda) [inline distribution/xfer/download.go:94]:
```go
type AddressableDownloadDescriptor
  DownloadDescriptor
  ...
}
```

**Comment by cpuguy83** [comment]
Outside of being able to make use of the growing list of containerd snapshotters, I'm not sure what the benefit is of adding containerd as a graph driver.

I was thinking we would just do a hard cutover to containerd that is opt-in. This would use the containerd content store, image store, snapshotters, and distribution code.

Discussed this briefly with some other maintainers a couple of weeks ago and thought we would just add a flag to enable the containerd integration, or the user would set the storage driver to something like `containerd-overlay` and that would use the containerd snapshotter, and this would only be the default setting, could potentially be overwritten with an option per HTTP request.
The main problem with this approach is it should really be an all or nothing... so not just a snapshotter, but also the content store, image store, etc.

Just using containerd as a graphdriver doesn't really solve any of the problems we have with the existing architecture.

Note that Docker already has some support for graphdriver plugins that could be used with your containerd integration out of tree.

**Comment by AkihiroSuda** [comment]
@cpuguy83 :+1: for per-container snapshotter opt, but it can be another PR

**Comment by cpuguy83** [comment]
That's just a side comment. My comment is more to the design of trying to cram snapshotters into the graphdriver interface.

**Comment by AkihiroSuda** [comment]
> I was thinking we would just do a hard cutover to containerd that is opt-in. This would use the containerd content store, image store, snapshotters, and distribution code.

Changing the distribution code without breaking compatibility is hard, so I think we should prioritize this PR over the distribution code.

**Comment by cpuguy83** [comment]
@AkihiroSuda What I'm saying here is, what problem are we solving by adding this. As far as I can tell it is adding more code to a legacy system.

**Comment by ktock** [comment]
@cpuguy83 @AkihiroSuda Thanks for the comments! 
The scope of this PR is narrower than the refactoring of cutover to containerd. This PR aims to use advanced features growing in containerd snapshotters (including "remote" snapshotters) also on Docker. I don't stick to the design of cramming snapshotters into graphdriver interface but this doesn't break API compatibilities and I believe this is easier to realize than the hard cutover.

> Note that Docker already has some support for graphdriver plugins that could be used with your containerd integration out of tree.

Yes, I know. Even in this case, "remote graphdriver" support will be needed. I can open the separated PR for it if needed.

**Comment by ktock** [comment]
Could I get more comments on this? Is this design unacceptable?

**Comment by AkihiroSuda** [comment]
@dmcgowan @tonistiigi @thaJeztah @cpuguy83 PTAL

**Comment by cpuguy83** [comment]
To be clear:

I don't think this is bad code. I think it's throwing new code into a problematic interface as a shortcut to the real end goal of replacing internal bits with containerd.
Because of that, I'd prefer not to bring this code in.

**Comment by thaJeztah** [comment]
I share @cpuguy83's concerns. I really (really!) would love to see the integration with containerd snapshotters and image-store, but am concerned that adding this feature would add additional complexity, and we would be painting ourself in a corner if we want to make that migration.

**Comment by ktock** [comment]
Thanks for clear feedbacks!:smile: So having an in-tree containerd-based graphdriver seems doesn't follow the vision towards containerd integration.

Is it even unacceptable to add new "features" (not graphdriver implementations) for distribution and graphdriver logic until the containerd integration completion? "Remote graphdriver" part in this PR is for tackling the issue "minimizing time to take for pull operation" and can be discussed separately from containerd integration. This proposes a backwards-compatible extension for distribution logic (e.g. interacting with graphdrivers and skipping downloading some layers) rather than having an additional graphdriver.

**Comment by AkihiroSuda** [comment]
Any idea to move this forward without breaking existing installations?

**Comment by neersighted** [comment]
Closed in favor of the new containerd migration work
